### DISCUSSION TITLE
Vector3 alignment and the NativeVector3 struct

### DISCUSSION THREAD

**Discussion by Halen84** [scripthookvdotnet/scripthookvdotnet#1261] [Q&A] [answered]
Hey, I'm the maintainer of the SHRDR2DN V2 (https://github.com/Halen84/ScriptHookRDR2DotNet-V2), and I had a few questions about how the `Vector3` and `NativeVector3` structs are handled and interpreted.

Why do you have an entire separate struct for `Vector3` in which the variable alignment is 8 bytes as opposed to the 4 bytes in `Vector3`? Why couldn't we just change `Vector3` alignment to 8 bytes and get rid of `NativeVector3` entirely? I've tested this on my repo and everything seems to work fine.

I understand that `NativeVector3` is for natives that return `Vector3` pointers, but this isn't necessary if the `Vector3` alignment was 8 to begin with.

I also understand that we're dealing with 2 different versions of the repo and 2 different versions of a game engine, so let me know if there's a specific reason on why we haven't already done this.

I've been in the process of updating the internals of my repo to be more up-to-date with this repo, since my repo is a descendant of this one.

Thanks.

**Comment by kagikn** [comment, marked as answer]
Hi, I know of what you worked in the SHRDR2DN V2 a bit. I'll talk about how much compatibility concerns us and how much we can reuse our codebase.

## Short Answer
For `Vector3`, we can't change the alignment in favor of compatibility as it can be used as a pointer type. For our codebase, maybe worth considering merging our codebase, but must be done with care before we do this.

## Long answer
We keep both `Vector3` and `NativeVector3` and the alignment of `Vector3` in favor of compatibility for existing scripts (the alignment is the main concern). You can read/write 4-byte aligned Vector3 at an arbitrary address as the layout is not set to `Auto`.

We really don't want to make breaking changes casually unless few script would face the changes or in major version upgrade (See [SemVer](https://semver.org/) to understand how we give versions). Actually, I innocently made breaking change in v2 API in dc09906 which affects since v2.10.11/v3.0.0 (the file was moved to another directory in 775e828), but that cause an observable disaster. It disabled some scripts to work as intended, like [this example](https://github.com/TORISOUP/GTAV_InfernoScripts/blob/eb7d498b7e18aa0590b3370eaffce8ddbbc03e2d/Inferno/InfernoScripts/Parupunte/Scripts/Owatashiki.cs#L17), where you are supposed to live with a very little health but dies in v3.0.0 or later SHVDN. Me realizing how the change caused too bad consequences: 😭 We can't really revert the change again though, since there's a lot of devs who want to keep using the v2 API even in 2023 (more than 3 years passed since v3 got available!)

As for the scripting runtime (`ScriptDomain`), it may be worth considering using an shared runtime, but we would have to deal with how to resolve API versions and how to detect the correct game engine (and behemoth `NativeMemory`). Also, I don't really expect that too much of the code would be able to be reused between different games, like [how crosire reacted in a new runtime for .NET Core](https://github.com/scripthookvdotnet/scripthookvdotnet/pull/1178#issuecomment-1454929396).

Actually I'm planning to release a new SHVDN variant built against CoreCLR that utilizes features only available in .NET Core/.NET 5+ such as interfaces with default implementations after the v4 API is released for this project, which will be different than [Sardelka9515's ScriptHookVDotNetCore](https://github.com/Sardelka9515/scripthookvdotnetcore). And also ones for RDR2. I haven't messed with RDR2 too much yet since I only have a PC where RDR2 runs but doesn't run too good even without a script mod (like at 50 fps with official minimum graphics  with the 1280 x 720 res?).

On the other hand, we are marking smelling interfaces where abstraction goes/went too wrong such as the setter of `World.RenderingCamera` while adding features. JayMontana36 said _why do SHVDN's "babying" wrappers deviate away_ in 5Mods Discord server. I agree that babying wrappers may confuse script devs so bad without careful API design if they abstract way more than native functions, and I like simpler but hard design better than easy but confusing (and probably complex) design. I also admit some interfaces added in earlier versions confuse script devs too badly, when crosire decided how API should be designed. Although JM gotta realize native functions often work as babying wrappers to some extent (he has to learn to reverse some natives to realize however).

Thank you for your understanding, I would say, but sounds worth continuing to discuss.

**Comment by Halen84** [comment]
Thanks for clearing this up, helped a lot. I'll be sure to NOT release a version where the `Vector3` alignment is 8 just for compatibility reasons (although I'm pretty sure I already destroyed all compatibility with older scripts already lol).

There's a LOT of differences between my repo and this one, which is part of the reason why I've been slow on merging them because I have no idea what will break, what's needed, what's not needed, if any bugs will emerge, if existing scripts will still work, etc. I've been monitoring commits to this repo usually just to see if its easily merge-able or even worth adding to mine. I've also been making my own little improvements that I think this repo could use, code related, but that's for another topic.

Since this repo and mine are almost the same thing, we could probably discuss changes to be made to both of the repos and collaborate. Again, this is for another topic.

As for https://github.com/scripthookvdotnet/scripthookvdotnet/pull/1178, I've known about this issue for quite a while, but it's slightly beyond my understanding to mess around with it myself. I've been trying to avoid making any big changes to `ScriptDomain.cs` because making it identical to this repo, just breaks a lot of things and needs lots of testing.

And yea, `NativeMemory` is pretty fucking huge and not easily merge-able, mainly just because of those function signatures. I've only bothered to look at them and not actually implement them lol. It's a real pain in the ass to have to go and find the RDR3 signatures so I've just left it be. I should probably get on that though.

Anyway, thanks for the help and info on all of this.

**Reply by kagikn** [reply]
I started finding signatures because I got tired of manually updating memory offsets that are changed too often, like ones on `CVehicle`. I started dumping exes since that happened, and my work got rewarded as a wider range of game version support with spending less time to update. Nightly builds would support almost all features in all game versions, except for ones that cannot be done due to absent of relevant game code in older game versions such as vehicle special flight stuff.

I don't like the decision how RPH and LSPDFR only support the latest version tbh. They don't want you to use them in older game versions so bad that they only publish the latest versions and unpublish old ones 🤡💀 LSPDFR has more forceful restriction iirc, since it does seem to connect somewhere online. If AB's ScriptHook supported only the latest version, we would have had to give up though.

**Reply by Halen84** [reply]
Luckily RDR2 never gets updated so function signatures being changed are the least of my concerns lol. I just gotta remap them.

**Comment by Halen84** [comment]
Btw if you ever need to contact me, my disc is `tuffytown`
